### DISCUSSION TITLE
[Packed Array] delta compression is almost useless for types smaller than or equal to 8bit

### DISCUSSION THREAD

**Discussion by AntonSulimenkoHarman** [ndsev/zserio#523] [Ideas]
I've prepared a packed array test with some delta compression-friendly patterns:
pattern | N | array | packed array
-- | -- | -- | --
0, 1, 2, 3, … | 4096 | 4098 | 4099
+1 ,-1, +1, … | 4096 | 4098 | 4099
+1,+1, -1,-1, ... | 4096 | 4098 | 4099
+1,+1,…,+1, -1,...-1,-1-1 1k groups | 4096 | 4098 | 4099
128, 128, 128, … | 4096 | 4098 | 4

ALL patterns (except a single case where the whole array is filled with a single const value) are larger because deltas are stored as 8bit!

One might say that delta compression prepares data for further archiving. So let's check it.
warning: !!! I've archived files with 7z, so sizes include service information like header and file names inside an archive. Please compare only the difference.
It's important to note that various archiving algorithms may produce varying results, but I am confident that this table accurately represents the concept.
pattern | array | packed array | diff
-- | -- | -- | --
0, 1, 2, 3, … | 421 | 425 | 4
+1 ,-1, +1, … | 192 | 198 | 6
+1,+1, -1,-1, ... | 201 | 208 | 7
+1,+1,…,+1, -1,...-1,-1-1 1k groups | 215 | 200 | -15
128, 128, 128, … | 190 | 166 | -24

Even in the case of further archiving, most of the patterns are better archivable as plain arrays by archiving algorithms.
And most favorable for delta compression patterns give only 24bytes and 15bytes advantage for 4K source data.

IMHO, delta compression is highly dependent on data patterns, and it becomes even more demanding when dealing with data that has a size of 8 bits or less.
Only data with large groups of the same data could give small benefits.
This makes it almost useless for real-world data.

**Comment by mikir** [comment]
I am afraid that I don't understand the first table. What does it mean column array and packed array? And which zserio type did you use?

In all cases, I am pretty sure that deltas are not stored as 8bit at all and they work. Please have a look e.g. [here](http://zserio.org/doc/ZserioInvisibles.html#packed-arrays-of-integers).

I agree that delta compression highly depend on data patterns but this is not a problem of zserio. Zserio just implements delta compression and cannot influence the data at all.

**Comment by fklebert** [comment]
@AntonSulimenkoHarman, could you provide a demo schema which you would intend to use with the data patterns you provide so that we can try out your assumptions?

Using general purpose compressors on zserio encoded structures is indeed a little bit of its own science. It is not directly the packed arrays that spoil the success there, but moreover any non-byte aligned schema. General purpose compressors usually work best on byte-aligned data patterns. So in case you have lists with e.g. `bit:5` or `bit:13` entries, it will almost always end up in generating non-compressable bytes in the encoded zserio stream, even if you add compression friendly value patterns.

**Comment by AntonSulimenkoHarman** [comment]
I apologize, I seem to have forgotten the schema.

```zs
struct Array<T> {
  varsize size;
  T data[size];
};

struct PackedArray<T> {
  varsize size;
  packed T data[size];
};

instantiate Array<uint8> ArrayUint8;
instantiate PackedArray<uint8> PackedArrayUint8;
```

Table description:
`N` - number of `data` (uint8) elements.
`array` - byte count of serialized `ArrayUint8` type.
`packed array` - byte count of serialized `PackedArrayUint8` type.

Warning!!! Just my assumption:
For me, `PackedArrayUint8` looks similar to an ASCII string.
Zserio shows a warning about the string: `'string' is not packable element type. [unpackable-array]`
So zserio somehow decides not to pack strings but packs string-like arrays.
```zs
instantiate PackedArray<string> PackedArrayString; //<<< warns about [unpackable-array]
instantiate PackedArray<PackedArray<uint8>> PackedArrayPackedArrayUint8; //<<< doesn't warn about [unpackable-array]
```
So it looks like zserio makes an assumption about string content (unlikely to be packable) but expects that PackedArrayUint8 can be packed.

BTW, does zserio effectively pack `bit:5` entries?

**Reply by fklebert** [reply]
As you are using a `uint8` array, how did you build the 4096 entries of the pattern 1,2,3,... exactly?
In case you had [..., 254, 255, 0, 1, 2, 3, ...] at position 255 in your generated pattern (so basically repeating the pattern once you hit maximum value of the underlying type), then this would definitely explain why you are not seeing expected results.
zserio is calculating the needed bits for storing the deltas from the content by checking for the biggest delta in the actual content. In this case the biggest offset would be 255, i.e. the complete range of `uint8` which effectively renders the delta encoding useless here as all entries are now stored with `uint8` width.

**Reply by fklebert** [reply]
> Warning!!! Just my assumption:
> For me, PackedArrayUint8 looks similar to an ASCII string.
> Zserio shows a warning about the string: 'string' is not packable element type. [unpackable-array]
> So zserio somehow decides not to pack strings but packs string-like arrays.

A zserio `string` is basically an `uint8` array with a size descriptor in front. That is correct.
But the assumption that a packed `uint8` array is the same as a packed `string` array is wrong as a string array is a 2-dimensional uint8 array in that regards.
And the packing of an array of an array does not work like the packing of an array itself.

**Comment by mikir** [comment]
> BTW, does zserio effectively pack `bit:5` entries?

Yes, zserio uses delta compression for all integer types, including bit:5 as well.

**Comment by fklebert** [comment]
@AntonSulimenkoHarman , @mikir 
I have hacked a small program that demonstrates the packing of the lists discussed above, so that we have something non-conceptional to discuss. Feel free to extend, comment, etc...
https://github.com/fklebert/zserio-experiments

**Reply by mikir** [reply]
It works properly here:

```
Unpacked Array: 
====================
size: 2056
bits per element: 8

Packed Array: 
====================
size: 539
bits per element: 2

Native Packed Array: 
====================
size: 539
bits per element: 2

Packed Array (3 times 0..255): 
====================
size: 6137
bits per element: 8
```

**Comment by AntonSulimenkoHarman** [comment]
Please check my test pattern extension: https://github.com/fklebert/zserio-experiments/pull/1

There are some unexpected (from the user's point of view) changes in the results for some patterns.
e.g. 0..255 + single zero
or 0,64,0,64,0,64....
or 0,0,0,0,0,0,0,0,0*N + single 64.

As I can see, please correct me if I'm wrong; a packed array of uint8 could be guaranteed to be effective (in terms of serialized data size) if the maximum diff between adjacent elements <= 5bit LSB.

**Reply by mikir** [reply]
I think these unexpected results become clearer as soon as users relealize that `packed`arrays in zserio are no magic, they are just packed using simple [delta encoding](https://en.wikipedia.org/wiki/Delta_encoding).

> As I can see, please correct me if I'm wrong; a packed array of uint8 could be guaranteed to be effective (in terms of serialized data size) if the maximum diff between adjacent elements <= 5bit LSB.

Depens on what one thinks is effective, but, yes, 5bit could be already quite ok. In general, the smaller maximum diff, the better compression.

**Reply by AntonSulimenkoHarman** [reply]
Completely agree. `packed` arrays are not a silver bullet and should be used carefully.
Some concluding remarks:
- 'packed' friendly data is data with the smallest MAXIMUM difference between EACH element in an array. Best pattern: sorted data without big steps. Sorted data with just one big step makes 'packed' array useless (larger size, slower runtime)
- 'packed' it's rather a data property, not a schema property. It's hard to guarantee packed-friendly data just by schema itself.
- It's hard to achieve 'packed' size effectivity in templates because it could be useful for one type and completely useless for another one.
- Each usage of 'packed' should be proven and not used blindly without reason.

**Reply by fklebert** [reply]
Let me disagree on these points and tell you why:
`packed` is implemented in a way that the array is never bigger than one 1 bit compared to store it non-packed. The zserio encoder works in a way that it checks whether it is better to pack or not to pack an array. Therefore it stores 1 bit in front of the array, telling the decoder whether the array is packed or not. So in worst case the array is only 1 bit bigger. This behavior is similar to zlib or other compressors. They also have a worst-case usage model that will generate bigger zipped streams than the original stream, also based on the content.

So in case you see anything else than an array being one bit bigger as it if would have been stored without the `packed` keyword, I would consider this as a bug of the implementation.

So using packed on lists is by design always a good idea if implemented correctly. Same as zipping content with a generic compressor.
We can of course discuss whether the current implementation of nested packed choices etc. is sufficiently done, but in general the concept is a good one in my opinion.

There is actually one small catch, which may hurt the generic compression on top: By adding that one bit in front of e.g. a string array, makes this string array nearly non-compressable as the strings are not byte-aligned any longer, giving generic compressors a hard time doing their job.
@mikir, this is something we should look into. Maybe align to the next byte before we actually put the data in. This adds 7 bits on every packed array, but should definitely help.
